src/raw/unix.rs

```diff
@@ -337,7 +359,54 @@ impl Socket {
         }
     }
 
-    ///Accepts incoming connection.
+    ///Accept a new incoming client connection and return its files descriptor and address.
+    ///
+    ///By default the newly created socket will be inheritable by child processes and created
+    ///in blocking I/O mode. This behaviour can be customized using the `flags` parameter:
+    ///
+    /// * `AcceptFlags::NON_BLOCKING`    – Mark the newly created socket as non-blocking
+    /// * `AcceptFlags::NON_INHERITABLE` – Mark the newly created socket as not inheritable by client processes
+    ///
+    ///Depending on the operating system's availablility of the `accept4(2)` system call this call
+    ///either pass the flags on to the operating system or emulate the call using `accept(2)`.
+    pub fn accept4(&self, flags: c_int) -> io::Result<(Socket, net::SocketAddr)> {
+        #[cfg(any(target_os = "linux", target_os = "freebsd", target_os = "openbsd", target_os = "dragonflybsd"))]
+        unsafe {
+            let mut storage: sockaddr_storage = mem::zeroed();
+            let mut len = mem::size_of_val(&storage) as socklen_t;
+
+            match accept4(self.inner, &mut storage as *mut _ as *mut _, &mut len, flags) {
+                SOCKET_ERROR => Err(io::Error::last_os_error()),
+                sock @ _ => {
+                    let addr = sockaddr_to_addr(&storage, len)?;
+                    Ok((Socket { inner: sock, }, addr))
+                }
+            }
+        }
+        
+        #[cfg(not(any(target_os = "linux", target_os = "freebsd", target_os = "openbsd", target_os = "dragonflybsd")))]
+        {
+            if (flags & !(libc::SOCK_NONBLOCK | libc::SOCK_CLOEXEC)) != 0 {
+                // `accept4()` also returns `EINVAL` if it receives an unknown flag
+                return Err(io::Error::from_raw_os_error(libc::EINVAL));
+            }
+
+            self.accept().map(|(sock, addr)| {
+                // Emulate the two most common (and useful) `accept4` flags using `ioctl`/`fcntl`
+                sock.set_nonblocking((flags & libc::SOCK_NONBLOCK) != 0).expect("Setting newly obtained client socket blocking mode");

```

Comment:
Returning that blob would be truely ugly indeed…

Reading the man-pages of the involved `ioctl` and `fcntl` system calls it appears the only things that can go wrong with these particular operations are either (a) programming errors on our side [meaning that we can and *should* panic in this case] or (b) another thread that randomly closes/modifies file descriptors that don't belong to it [not a problem we can or should care about]. We can be confident about this, because we never pass of the file descriptor to code we don't control until *after* we complete all operations. Since there is no reasonable way for these calls to fail (aside from misbehaving threads or operating system kernels), I'd strongly recommend to keep this as-is.
Better documentation of this fact would be in order through, I agree on that…